File: stage1/ClassLoaderCache.scala
package cbt

import java.net._
import java.util._
import collection.JavaConverters._

case class ClassLoaderCache(
  logger: Logger,
  private[cbt] hashMap: java.util.Map[AnyRef,AnyRef]
){
  val cache = new KeyLockedLazyCache[ClassLoader]( hashMap, Some(logger) )
  override def toString = (
    s"ClassLoaderCache("
    ++
    hashMap.asScala.collect{
      case (key, value) if key.isInstanceOf[String] =>
        key.toString.split(":").mkString("\n") -> value
    }.toVector.sortBy(_._1).map{
      case (key, value) => key + " -> " + hashMap.get(value)
    }.mkString("\n\n","\n\n","\n\n")
    ++
    ")"
  )
}

File: nailgun_launcher/ClassLoaderCache.java
package cbt;

import java.util.*;
import static java.io.File.pathSeparator;
import static cbt.Stage0Lib.*;

final public class ClassLoaderCache{
  public Map<Object,Object> hashMap;
  final ThreadLocal<HashSet<String>> seen = new ThreadLocal<HashSet<String>>(){
    @Override protected HashSet<String> initialValue(){
      return new HashSet<String>();
    }
  };

  public ClassLoaderCache(
    Map<Object,Object> hashMap
  ){
    this.hashMap = hashMap;
  }

  public ClassLoader get( String key, long timestamp ){
    seen.get().add( key );
    @SuppressWarnings("unchecked")
    ClassLoader t = (ClassLoader) hashMap.get(
      hashMap.get( key )
    );
    assert hashMap.get(t).equals(timestamp);
    return t;
  }

  public boolean containsKey( String key, long timestamp ){
    boolean contains = hashMap.containsKey( key );
    if( contains ){
      Object keyObject = hashMap.get( key );
      Object classLoader = hashMap.get( keyObject );
      long oldTimestamp = (long) hashMap.get( classLoader );
      boolean res = oldTimestamp == timestamp;
      return res;
    } else {
      return false;
    }
  }

  public void put( String key, ClassLoader value, long timestamp ){
    assert !seen.get().contains( key ): "Thread tries to update cache key after observing it: " + key;
    LockableJavaKey keyObject = new LockableJavaKey();
    hashMap.put( key, keyObject );
    hashMap.put( keyObject, value );
    hashMap.put( value, timestamp );
  }

  @Override public String toString(){
    StringBuilder res = new StringBuilder();
    res.append("ClassLoaderCache(\n\n");
    for( Object key: hashMap.keySet() ){
      if( key instanceof String )
        res.append(
          mkString( "\n", key.toString().split(":") ) + " -> " + hashMap.get( hashMap.get(key) )
          + "\n\n"
        );
    }
    res.append("\n\n");
    return res.toString();
  }
}
class LockableJavaKey{}
